### PR TITLE
fix: Determine default async concurrency on a per-model basis

### LINKED ISSUE #2032: [bug][evals] Gemini with high concurrency causes concurrency issues
I'm getting a host of errors with Gemini.
Notebook
https://colab.research.google.com/drive/1JlU0hL04RbC7HaVDf_xszvhy2ARdOXJI?usp=sharing

WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
WARNING:urllib3.connectionpool:Connection pool is full, discarding connection: metadata.google.internal. Connection pool size: 10
Exception in worker on attempt 1: raised ValueError('Content has no parts.')
Requeuing...
Exception in worker on attempt 1: raised ValueError('Content has no parts.')
Requeuing...
Exception in worker on attempt 1: raised ValueError('Content has no parts.')
Requeuing...
Exception in worker on attempt 1: raised ValueError('Content has no parts.')

------------------------------

tion in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised AttributeError("'list' object has no attribute 'lower'")                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...                                                                                                                                  
Exception in worker on attempt 3: raised InvalidArgument('Request contains an invalid argument.')                                             
Requeuing...

### LINKED ISSUE #2084: Adjust default concurrencies for llm evals to mitigate concurrency issues across vendors

### PR DESCRIPTION
**Pull request by anticorrelator**
resolves #2032 
resolves #2084 

Some of the LLM model SDKs that we wrap do not always work well when used at high concurrency, this PR adds a mechanism to set the default on a per-vendor basis.

### CODE REVIEW

File: src/phoenix/experimental/evals/functions/classify.py [hunk 1]
```diff
@@ -116,8 +116,8 @@ def llm_classify(
         run_sync (bool, default=False): If True, forces synchronous request submission. Otherwise
         evaluations will be run asynchronously if possible.
 
-        concurrency (int, default=20): The number of concurrent evals if async submission is
-        possible.
+        concurrency (Optional[int], default=None): The number of concurrent evals if async
+        submission is possible. If not set, the default is determined on a per-model basis.
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/classify.py:120]:
```suggestion
        concurrency (Optional[int], default=None): The number of concurrent evals if async
        submission is possible. If not provided, a recommended default concurrency is set on a per-model basis.
```

Suggestion: Make it clear that the default value is the recommended value.

File: src/phoenix/experimental/evals/functions/classify.py [hunk 2]
```diff
@@ -127,6 +127,8 @@ def llm_classify(
         from the entries in the rails argument or "NOT_PARSABLE" if the model's output could
         not be parsed.
     """
+    # read the default concurrency from the model if not specified
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/classify.py:130]: I don't think this comment is needed.

File: src/phoenix/experimental/evals/functions/classify.py [hunk 3]
```diff
@@ -381,13 +383,20 @@ def run_evals(
         as model invocation parameters and details about retries and snapping to
         rails.
 
-        concurrency (int, optional): The number of concurrent evals if async
-        submission is possible.
+        concurrency (Optional[int], default=None): The number of concurrent evals if async
+        submission is possible. If not set, the default is determined on a per-model basis.
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/classify.py:387]: Same suggestion as above.

File: src/phoenix/experimental/evals/functions/generate.py [hunk 4]
```diff
@@ -70,14 +70,17 @@ def llm_generate(
         run_sync (bool, default=False): If True, forces synchronous request submission. Otherwise
         evaluations will be run asynchronously if possible.
 
-        concurrency (int, default=20): The number of concurrent evals if async submission is
-        possible.
+        concurrency (int, optional): The number of concurrent evals if async
+        submission is possible.
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/generate.py:74]: I would use the same docstring as above.

File: src/phoenix/experimental/evals/functions/generate.py [hunk 5]
```diff
@@ -70,14 +70,17 @@ def llm_generate(
...
 
     Returns:
         generations_dataframe (pandas.DataFrame): A dataframe where each row
         represents the generated output
 
     """
+    # read the default concurrency from the model if not specified
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/generate.py:81]: Same as above.

File: src/phoenix/experimental/evals/functions/classify.py [hunk 6]
```diff
@@ -381,13 +383,20 @@ def run_evals(
...
     Returns:
         List[DataFrame]: A list of dataframes, one for each evaluator, all of
         which have the same number of rows as the input dataframe.
     """
+    # use the minimum default concurrency of all the models
+    if concurrency is None:
+        if len(evaluators) == 0:
+            concurrency = 1
+        else:
+            concurrency = min(evaluator.default_concurrency for evaluator in evaluators)
```
Comment (axiomofjoy) [inline src/phoenix/experimental/evals/functions/classify.py:398]:
```suggestion
    if not evaluators:
        return []

    concurrency = concurrency or min(evaluator.default_concurrency for evaluator in evaluators)
```

**Review by axiomofjoy — APPROVED** [review]
Nice job! Don't forget to update the documentation.

File: src/phoenix/experimental/evals/functions/classify.py (hunk 1 above)
Comment (axiomofjoy) [reply src/phoenix/experimental/evals/functions/classify.py:120]: This makes me wonder if we really should allow the user to use many different models inside of `run_evals`.

File: src/phoenix/experimental/evals/functions/classify.py (hunk 6 above)
Comment (axiomofjoy) [reply src/phoenix/experimental/evals/functions/classify.py:398]: This makes me wonder if we really should allow the user to use multiple different models inside `run_evals`.

Comment (anticorrelator) [reply src/phoenix/experimental/evals/functions/classify.py:398]: I think it's fine so far? I think this is a pretty small concession that we can probably work around

**Pull request merged**